mhs-reference-implementation/mhs/handler/client_request_handler.py

```diff
@@ -1,31 +1,81 @@
 import logging
+from datetime import timedelta
+from typing import Callable, Dict
 
+from tornado.locks import Event
 from tornado.web import RequestHandler, HTTPError
 
-from mhs.sender.sender import UnknownInteractionError
+from mhs.sender.sender import UnknownInteractionError, Sender
 
 
 class ClientRequestHandler(RequestHandler):
     """A RequestHandler for client requests to this MHS."""
 
-    def initialize(self, sender):
+    def initialize(self, sender: Sender, callbacks: Dict[str, Callable[[str], None]], async_timeout: int):
         """Initialise this request handler with the provided configuration values.
 
         :param sender: The sender to use to send messages.
+        :param callbacks: The dictionary of callbacks to use when awaiting an asynchronous response.
+        :param async_timeout: The amount of time (in seconds) to wait for an asynchronous response.
         """
         self.sender = sender
+        self.callbacks = callbacks
+        self.async_response_received = Event()
+        self.async_timeout = async_timeout
 
-    def post(self):
+    async def post(self):
         logging.debug("Client POST received: %s", self.request)
 
         interaction_name = self.request.uri[1:]
 
+        async_message_id = None
         try:
-            response = self.sender.send_message(interaction_name, self.request.body.decode())
+            async_message_id, message = self.sender.build_message(interaction_name, self.request.body.decode())
 
-            logging.debug("Message sent. Received response: %s", response)
+            interaction_is_asynchronous = async_message_id is not None
+            if interaction_is_asynchronous:
+                self.callbacks[async_message_id] = self._write_async_response
+                logging.debug("Added callback for asynchronous message with ID '%s'", async_message_id)
+
+            immediate_response = self.sender.send_message(interaction_name, message)
+
+            if interaction_is_asynchronous:
```

Comment:
This if and the if above are the same? Is there a reason for this.

Reply:
I wanted to store the callback (only if the message is asynchronous) before sending any requests (which is done for synchronous and asynchronous messages). The second if statement then decides whether to wait for an asynchronous response or return immediately.

Reply:
As discussed offline. Wondering if the immediate_response we get back from `spine` can be used to determine whether it is the actual response of just an acknowledgment and using this to determine whether to pause or send the actual response. Could be one to investigate later.

Reply:
That's a good call. I'll add this to the PoC documentation around how we're going to expand it for more messaging patterns.